source/common/event/scaled_range_timer.cc

```diff
@@ -0,0 +1,220 @@
+    const ScaledRangeTimerManager::BucketEnabledList::iterator bucket_position;
+  };
+
+  void onPendingTimerComplete() {
+    ENVOY_LOG_MISC(info, "pending complete for {}", static_cast<void*>(this));
+    ASSERT(absl::holds_alternative<WaitingForMin>(state_));
+    WaitingForMin& pending = absl::get<WaitingForMin>(state_);
+
+    const MonotonicTime now = manager_.dispatcher_.approximateMonotonicTime();
+    auto it = manager_.add(*this, pending.latest_trigger - now);
+    state_.emplace<ScalingMax>(pending.latest_trigger - now, it);
+  }
+
+  ScaledRangeTimerManager& manager_;
+  const TimerCb callback_;
+  const TimerPtr pending_timer_;
+
+  absl::variant<Inactive, WaitingForMin, ScalingMax> state_;
+  const ScopeTrackedObject* scope_;
+};
+
+ScaledRangeTimerManager::ScaledRangeTimerManager(Dispatcher& dispatcher, float scale_factor,
+                                                 std::chrono::milliseconds minimum_duration)
+    : dispatcher_(dispatcher), minimum_duration_(minimum_duration), scale_factor_(scale_factor) {}
+
+RangeTimerPtr ScaledRangeTimerManager::createTimer(TimerCb callback) {
+  return std::make_unique<RangeTimerImpl>(callback, *this);
+}
+
+void ScaledRangeTimerManager::setScaleFactor(float scale_factor) {
+  scale_factor_ = DurationScaleFactor(scale_factor);
+  for (auto& bucket : buckets_) {
+    bucket.updateTimer(*this, true);
+  }
+}
+
+MonotonicTime::duration
+ScaledRangeTimerManager::getBucketedDuration(MonotonicTime::duration max_duration) {
+  if (max_duration <= MonotonicTime::duration::zero()) {
+    return MonotonicTime::duration::zero();
+  }
+  const int index = std::floor(std::log(max_duration.count()) / std::log(kBucketScaleFactor));
+  return MonotonicTime::duration(
+      static_cast<MonotonicTime::duration::rep>(std::pow(kBucketScaleFactor, index)));
+}
+
+void ScaledRangeTimerManager::Bucket::updateTimer(ScaledRangeTimerManager& manager,
+                                                  bool scale_factor_changed) {
+  if (scaled_timers.empty()) {
+    if (timer->enabled()) {
+      timer->disableTimer();
+    }
+  } else { // !scaled_timers.empty()
+    if (scale_factor_changed || !timer->enabled()) {
+      auto& entry = scaled_timers.front();
+      timer->enableTimer(std::chrono::duration_cast<std::chrono::milliseconds>(
+          manager.scale_factor_.value() *
+          (entry.latest_trigger_time - manager.dispatcher_.approximateMonotonicTime())));
+    }
+  }
+}
+
+ScaledRangeTimerManager::DurationScaleFactor::DurationScaleFactor(float value)
+    : value_(std::max(0.0f, std::min(value, 1.0f))) {}
+
+float ScaledRangeTimerManager::DurationScaleFactor::value() const { return value_; }
+
+ScaledRangeTimerManager::BucketEnabledList::iterator
+ScaledRangeTimerManager::add(RangeTimerImpl& timer, const MonotonicTime::duration max_duration) {
+  Bucket& bucket = getOrCreateBucket(max_duration);
+  const auto quantized_duration = getBucketedDuration(max_duration);
```

Comment:
What guarantees are you providing with your implementation?  I thought we wanted to prefer having timers fire close to the configured max time when the overload factor is 0.0, scale factor is 1.0

Reply:
Yeah, this is something I was hoping to hash out here on the PR. Here are a couple options, most of which we've discussed
1. [as implemented] adjust max duration based on bucket so the timer fires on time for S=0 and early for S=1, but within the window
1. adjust min duration for the wait-for-min phase so that `max-min` is a bucket boundary; this will fire late for S=0 but within the window, and on-time for S=1
1. switch between these two by looking at the current scaling factor when enableTimer is called; this works really well if the scale factor doesn't change too much
1. use more buckets by decreasing the exponentiation factor

I'm a fan of option 2 over 1 since it is perfectly accurate for S=1, which preserves the existing behavior for proxies that don't change the default. For S=0, the timer might fire anywhere between `min` and `min + (max - min)/2`. We can do better than that by adding in option 4.

I don't know what to think about option 3. It seems like an extra complication that relies on heuristics for correctness. Maybe worth implementing later but I don't think we need it.

@antoniovicente thoughts? If this seems okay I'll go ahead switching from 1 to 2 and reducing the exponentiation factor.

Reply:
I'm not seeing the following option among the ones listed above:
Add timers based on max_configured-min_configured to a bucket.  Use a priority queue or multiset to keep the timers sorted by max expiry time.  Expire the timer with the smallest max expiry time using the following formula:
  expiry_time - overload_factor * (max_configured-min_configured) <= current_time

Properties of the algorithm above include: timers fire at the right time for S=1.  Also, timers fire at the right time when scaling if all entries in the bucket have the same (max-min).

I'm not sure how to get other people to look over these proposals.  I think that I'm currently the only person reviewing this PR.  It may be necessary to move the conversation back to the associated issue or a design doc.